source/controllers/asd.py

```diff
@@ -0,0 +1,162 @@
+# limitations under the License.
+
+"""
+This module contains the asd controller (logic for managing asds)
+"""
+import os
+import re
+import json
+import math
+import random
+import signal
+import string
+import datetime
+from subprocess import check_output
+from source.tools.fstab import FSTab
+from source.tools.services.service import ServiceManager
+from source.tools.configuration import EtcdConfiguration
+from source.tools.localclient import LocalClient
+
+local_client = LocalClient()
+
+
+class ASDController(object):
+    NODE_ID = os.environ['ASD_NODE_ID']
+    ASD_SERVICE_PREFIX = 'alba-asd-'
+    ASD_CONFIG_ROOT = '/ovs/alba/asds/{0}'
+    ASD_CONFIG = '/ovs/alba/asds/{0}/config'
+    ASDS = {}
+    CONFIG_ROOT = '/ovs/alba/asdnodes/{0}/config'.format(NODE_ID)
+
+    @staticmethod
+    def _log(message):
+        print '{0} - {1}'.format(str(datetime.datetime.now()), message)
+
+    @staticmethod
+    def list_asds(mountpoint):
+        asds = {}
+        try:
+            for asd_id in os.listdir(mountpoint):
+                if os.path.isdir('/'.join([mountpoint, asd_id])) and EtcdConfiguration.exists(ASDController.ASD_CONFIG.format(asd_id)):
+                    asds[asd_id] = EtcdConfiguration.get(ASDController.ASD_CONFIG.format(asd_id))
+                    output = check_output('ls {0}/{1}/ 2>&1 || true'.format(mountpoint, asd_id), shell=True)
+                    if 'Input/output error' in output:
+                        asds[asd_id]['state'] = {'state': 'error',
+                                                 'detail': 'ioerror'}
+                    else:
+                        service_name = '{0}{1}'.format(ASDController.ASD_SERVICE_PREFIX, asd_id)
+                        if ServiceManager.has_service(service_name, local_client):
+                            service_state = ServiceManager.get_service_status(service_name, local_client)
+                            if service_state is False:
+                                asds[asd_id]['state'] = {'state': 'error',
+                                                         'detail': 'servicefailure'}
+                            else:
+                                asds[asd_id]['state'] = {'state': 'ok'}
+                        else:
+                            asds[asd_id]['state'] = {'state': 'error',
+                                                     'detail': 'servicefailure'}
+        except OSError as ex:
+            ASDController._log('Error collecting ASD information: {0}'.format(str(ex)))
+        return asds
+
+    @staticmethod
+    def create_asd(disk):
+        all_asds = {}
+        mountpoints = FSTab.read()
+        for _disk, mountpoint in mountpoints.iteritems():
+            all_asds.update(ASDController.list_asds(mountpoint))
+        mountpoint = mountpoints[disk]
+
+        # Fetch disk information
+        df_info = check_output('df -k {0} || true'.format(mountpoint), shell=True).strip()
```

Comment:
Doesn't ```df``` has an option to only print the required field and also in the correct sizing?